CONTRACTS: Populate function name in source locations

nwetzler opened the pull request:
During parsing of function parameters, the function name should be set
in the source location, but only for top-level function declarations
and definitions.  In addition, function contracts should be only
allowed for top-level function declarations and definitions.

CBMC Viewer reports a failure when analyzing the incomplete coverage
report, causing the CMBC Starter Kit to fail during a standard build.

Also removed source location from "end function" command during
contract checking and added proper source location to function
contract instrumentation.

Resolves issues #7032, #7070

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [X] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig commented:
@nwetzler The failing tests still need looking at.

kroening commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -674,6 +674,28 @@ void c_typecheck_baset::apply_asm_label(
   }
 }
 
+/// Recursively set function name in source locations of expression
```
Would it not be better to set the function name when parsing the contract?

nwetzler replied:
Remi and I were under the impression that the function name doesn't exist or isn't readily available at parsing time.  This is because the function doesn't fully exist until the body has been parsed.  If it is possible during parsing, that would be a better solution, but I didn't see a way to make it happen.

kroening replied:
You've got the function name once the declarator has been seen; this is well before the contract.

nwetzler replied:
@kroening I've now changed this in the parser and have removed changes from type-checking.  The parser will now set the function in source location, but only for top-level functions.  I think this change results in the correct behavior and excludes problems introduced by function pointers.

nwetzler commented:
> @nwetzler The failing tests still need looking at.

I've updated the contract tests to include the function name in regular expressions.

Commit:
CONTRACTS: Populate function name in source locations

During parsing of function parameters, the function name should be set
in the source location, but only for top-level function declarations
and definitions.  In addition, function contracts should be only
allowed for top-level function declarations and definitions.

CBMC Viewer reports a failure when analyzing the incomplete coverage
report, causing the CMBC Starter Kit to fail during a standard build.

Also removed source location from "end function" command during
contract checking and added proper source location to function
contract instrumentation.

Resolves issues #7032, #7070

jimgrundy commented on src/ansi-c/parser.y:
```diff
@@ -3407,6 +3407,15 @@ cprover_function_contract_sequence_opt:
+          // be disallowed.
+          int contract_in_global_scope = (PARSER.scopes.size() == 1);
+          int contract_in_top_level_function_scope = (PARSER.scopes.size() == 2);
+          assert(contract_in_global_scope || contract_in_top_level_function_scope);
```
What happens if you say something like 
yyerror("function contracts allowed only at top-level declarations.")

nwetzler replied:
@jimgrundy Is this what you had in mind?
```
          int contract_in_global_scope = (PARSER.scopes.size() == 1);
          int contract_in_top_level_function_scope = (PARSER.scopes.size() == 2);
          if(!contract_in_global_scope && !contract_in_top_level_function_scope)
          {
            yyerror("Function contracts allowed only at top-level declarations.");
          }
          assert(contract_in_global_scope || contract_in_top_level_function_scope);
```

It results in an additional line of output.
```
Parsing main.c
file main.c line 8 function foo: Function contracts allowed only at top-level declarations. before ')'
Assertion failed: (contract_in_global_scope || contract_in_top_level_function_scope), function yyansi_cparse, file parser.y, line 3421.
Abort trap: 6
```

jimgrundy replied:
Almost. I think we should do that to get a better error message and then error out more gracefully than an assertion violation. I don't think there should be any way user input should cause an internal assertion to fail -- failure of internal assertions should be seen as evidence of a tool bug.

Parsers don't exit immediately on the first error either.  How about this instead, this is the function the lexer uses to report errors, but it is also available in the parser:
yyansi_cerror("Function contracts allowed only at top-level declarations.").

Maybe instead of the assertion violation we just set $$ to something empty so the bad thing never gets declared? There should be some way to keep going for now but record the fact the the parse isn't clean, maybe yyansi_cerror will do that?

nwetzler replied:
@jimgrundy I've updated the error handling to print debug information through `yyerror` and then abort cleanly with `YYABORT`.  This produces output like:
```
CBMC version 5.63.0 (cbmc-5.63.0-29-g6a8c9a673-dirty) 64-bit x86_64 macos
Parsing ../embedded_contract_fail_02/main.c
file ../embedded_contract_fail_02/main.c line 8 function foo: Function contracts allowed only at top-level declarations. before ')'
PARSING ERROR
```

feliperodri commented on regression/contracts/embedded_contract_fail_01/test.desc:
```diff
@@ -0,0 +1,12 @@
+--
+Checks if function contracts can be attached to function pointers
+(with non-empty parameter lists) in function parameters.  This should
+fail.  Exit code 64 for Windows servers.
```
You don't need to include "Exit code 64 for Windows servers." in the description, the error code itself is self-explanatory.

nwetzler replied:
I felt better reminding myself why there are two acceptable exit codes.  I'd like to keep the comment unless the 1 and 64 pattern is super common.

feliperodri replied:
The pattern is quite common as it appears in 118 test description, but there is no harm in leaving this reminder here as well.

jimgrundy reviewed (approved):
Thanks for those changes. I like it.

feliperodri commented on src/ansi-c/parser.y:
```diff
@@ -3407,6 +3407,19 @@ cprover_function_contract_sequence_opt:
+          // top-level function declaration or top-level function
+          // definition.  Any embedded function pointer scopes should
+          // be disallowed.
+          int contract_in_global_scope = (PARSER.scopes.size() == 1);
+          int contract_in_top_level_function_scope = (PARSER.scopes.size() == 2);
```
Could you explain the magic numbers `1` and `2` in the comment?

nwetzler replied:
I was hoping the long variable names, combined with the comment above, would be self-documenting.  When we initialize the parser, we start by pushing a global scope onto the stack.  (There's an invariant that we never have an empty scopes stack.)  So a size of 1 means the stack only contains the global scope.  A size of two means we must be in some top-level event like a function declaration or definition because we've pushed something else on the stack.

feliperodri commented on src/ansi-c/parser.y:
```diff
@@ -3447,15 +3460,41 @@ postfixing_abstract_declarator:
+            // Set function name (last declarator) in source location
+            // before parsing function contracts.  Only do this if we
+            // are at a global scope.
+            if (PARSER.current_scope().prefix.empty()) {
```
@kroening is there a better way to check if we are at a global scope?

nwetzler replied:
I borrowed this methodology from the `add_declarator` function in `ansi_c_parser.cpp`:
```
  // global?
  if(current_scope().prefix.empty())
    ansi_c_declaration.set_is_global(true);
```

I had originally written something like:
`if (PARSER.current_scope() == PARSER.root_scope())`
But there's no equality comparator defined for the scope class.

feliperodri reviewed (approved):
LGTM, only minor comments.

SaswatPadhi reviewed (approved):
Thanks! LGTM
